source/Grid/Grid.js

```diff
@@ -1111,19 +1129,49 @@ class Grid extends React.PureComponent<Props, State> {
 
     // Render only enough columns and rows to cover the visible area of the grid.
     if (height > 0 && width > 0) {
-      const visibleColumnIndices = instanceProps.columnSizeAndPositionManager.getVisibleCellRange(
+      let visibleColumnIndices = instanceProps.columnSizeAndPositionManager.getVisibleCellRange(
         {
           containerSize: width,
           offset: scrollLeft,
         },
       );
-      const visibleRowIndices = instanceProps.rowSizeAndPositionManager.getVisibleCellRange(
+      let visibleRowIndices = instanceProps.rowSizeAndPositionManager.getVisibleCellRange(
         {
           containerSize: height,
           offset: scrollTop,
         },
       );
 
+      // Compute max columns within visible area
+      if (this._maxRenderedColumnCount == 0) {
+        this._maxRenderedColumnCount =
+          visibleColumnIndices.stop - visibleColumnIndices.start;
+        this._maxRenderedColumnCount++;
+      }
+
+      if (
+        this._maxRenderedColumnCount !=
+        visibleColumnIndices.stop - visibleColumnIndices.start
+      ) {
+        if (visibleColumnIndices.start > 0) visibleColumnIndices.start--;
+        else visibleColumnIndices.stop++;
+      }
+
+      // Compute max rows within visible area
+      if (this._maxRenderedRowCount == 0) {
+        this._maxRenderedRowCount =
+          visibleRowIndices.stop - visibleRowIndices.start;
+        this._maxRenderedRowCount++;
+      }
+
+      if (
+        this._maxRenderedRowCount !=
+        visibleRowIndices.stop - visibleRowIndices.start
+      ) {
+        if (visibleRowIndices.start > 0) visibleRowIndices.start--;
+        else visibleRowIndices.stop++;
+      }
+
       const horizontalOffsetAdjustment = instanceProps.columnSizeAndPositionManager.getOffsetAdjustment(
```

Comment:
It would be cleaner if this code was encapsulated in a function instead of all present inline here.